File: 2018-12/2018-12-06/UnivalTreeCount.cpp
#include <iostream>

template<typename T>
struct Node
{
	int value;
	
	Node* left = nullptr;
	Node* right = nullptr;

	Node(int value, Node* left, Node* right)
		: value(value), left(left), right(right)
	{ }
};

template<typename T>
bool IsUnival(Node<T>* node, unsigned int& counter)
{
	if (node == nullptr)
	{
		return true;
	}

	bool isLeftUnival = IsUnival(node->left, counter);
	bool isRightUnival = IsUnival(node->right, counter);

	if (!isLeftUnival || !isRightUnival)
	{
		return false;
	}

	if ((node->left != nullptr) && (node->value != node->left->value))
	{
		return false;
	}

	if ((node->right != nullptr) && (node->value != node->right->value))
	{
		return false;
	}

	counter++;

	return true;
}

template<typename T>
unsigned int GetUnivalTreeCount(Node<T>* root)
{
	unsigned int count = 0;

	return (IsUnival(root, count), count);
}

template<typename T>
void FreeNodeMemory(Node<T>* root)
{
	if (root == nullptr)
	{
		return;
	}

	FreeNodeMemory(root->left);
	FreeNodeMemory(root->right);

	delete root;
	root = nullptr;
}

int main(int argc, char* argv[])
{
	Node<int>* grandGrandChildA = new Node<int>(1, nullptr, nullptr);
	Node<int>* grandGrandChildB = new Node<int>(1, nullptr, nullptr);

	Node<int>* grandChildA = new Node<int>(1, grandGrandChildA, grandGrandChildB);
	Node<int>* grandChildB = new Node<int>(0, nullptr, nullptr);

	Node<int>* childA = new Node<int>(1, nullptr, nullptr);
	Node<int>* childB = new Node<int>(0, grandChildA, grandChildB);

	Node<int>* root = new Node<int>(0, childA, childB);

	std::cout << GetUnivalTreeCount(root) << std::endl;

	FreeNodeMemory(root);

	std::cin.get();

	return 0;
}

File: Days 001 - 010/Day 08/UnivalTreeCount.cs
using System;

namespace DailyCodingProblem
{
	internal class Node
	{
		public int Value { get; private set; }

		public Node Left { get; private set; }
		public Node Right { get; private set; }

		public Node(int value = 0, Node left = null, Node right = null)
		{
			this.Value = value;
			this.Left = left;
			this.Right = right;
		}
	}

	internal class Day8
	{
		private static int Main(string[] args)
		{
			Node grandGrandChildA = new Node(1);
			Node grandGrandChildB = new Node(1);
			Node grandChildA = new Node(1, grandGrandChildA, grandGrandChildB);
			Node grandChildB = new Node(0);
			Node childA = new Node(1);
			Node childB = new Node(0, grandChildA, grandChildB);
			Node root = new Node(0, childA, childB);

			Console.WriteLine(GetUnivalTreeCount(root));

			Console.ReadLine();

			return 0;
		}

		private static int GetUnivalTreeCount(Node root)
		{
			int count = 0;
			IsTreeUnival(root, ref count);

			return count;
		}

		private static bool IsTreeUnival(Node node, ref int counter)
		{
			if (node == null)
			{
				return true;
			}

			bool isLeftUnival = IsTreeUnival(node.Left, ref counter);
			bool isRightUnival = IsTreeUnival(node.Right, ref counter);

			if (!isLeftUnival || !isRightUnival)
			{
				return false;
			}

			if (node.Left != null && node.Value != node.Left.Value)
			{
				return false;
			}

			if (node.Right != null && node.Value != node.Right.Value)
			{
				return false;
			}

			counter++;

			return true;
		}
	}
}